rfcs/0110-inherit-as-list.md

```diff
@@ -0,0 +1,153 @@
+---
+feature: inherit-as-list
+start-date: 2021-10-17
+author: Ryan Burns (@r-burns)
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+This RFC proposes a new Nix syntax `[ inherit (<attrset>) <attrnames>; ]`,
+which constructs a list from the values of an attrset.
+
+The goal is to provide a similarly-terse but more principled alternative
+to the often-used `with <attrset>; [ <attrnames> ]`.
+
+# Motivation
+[motivation]: #motivation
+
+It is currently cumbersome to create a list from the values of an attrset.
+If one has an attrset `attrs` and wishes to create a list containing some of
+its values, one could naively write:
+
+```nix
+[ attrs.a attrs.b attrs.c ]
+```
+
+To avoid typing `attrs` many times, one will typically use `with` instead:
+
+```nix
+with attrs; [ a b c ]
+```
+
+However, the `with` expression has many well-known drawbacks, such as
+unintuitive shadowing behavior [1][2], prevention of static scope checking [3][4],
+and reduced evaluation performance [3].
+
+* [1] https://github.com/NixOS/nix/issues/490
+* [2] https://github.com/NixOS/nix/issues/1361
+* [3] https://github.com/NixOS/nixpkgs/pull/101139
+* [4] https://nix.dev/anti-patterns/language#with-attrset-expression
+
+Nonetheless, Nix expression authors are subtly guided toward the `with` form
+because it is (or at least appears) simpler than any existing alternatives.
+Some alternatives are suggested in
+https://nix.dev/anti-patterns/language#with-attrset-expression, but as these
+are more verbose and complex than `with`, they are rarely used.
+
+The goal of this RFC is to provide a similarly-terse alternative which avoids
+these drawbacks.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposed syntax is:
+
+```
+[ inherit (attrs) a b c; ]
+[ inherit a b c; ]
+```
+
+These expressions are (respectively) syntactic sugar for:
+
+```nix
+builtins.attrValues { inherit (attrs) a b c; }
+builtins.attrValues { inherit a b c; }
```

Comment:
This seems like a suboptimal choice.

1. It removes the order from the attributes.
2. It is unclear how name collisions are handled.

Why not just make it desugar to 

```nix
[ attrs.a attrs.b attrs.c ]
[ a b c ]
```

Reply:
Ah, I had planned for `[inherit (foo) path; inherit (bar) path;]` to desugar to `(builtins.attrValues { inherit (foo) path; }) ++ (builtins.attrValues { inherit (bar) path; })` under the current draft of the proposal, performing desugaring independently on each `inherit` group. Also, duplicate attrs (e.g. `[ inherit (foo) path path; ]` would cause an error rather than deduplicate. But this is definitely not intuitive.

Your other points are well-taken, and I had not considered them. I think I'm coming around to your point of view on this... the `attrValues { inherit }` semantics are certainly not as simple as I thought!

Reply:
> performing desugaring independently on each inherit group

Ah, if the different attrs generate to different lists it is much better than I originally understood.

The sorting by keys is still seems a bit weird that given `[inherit a b; c d]`, `[inherit b a; c d]` is a noop but  `[inherit a b; d c]` changes the order but that is minor.

Reply:
> performing desugaring independently on each inherit group

Ah, if the different attrs generate to different lists it is much better than I originally understood.

The sorting by keys is still seems a bit weird that given `[inherit a b; c d]`, `[inherit b a; c d]` is a noop but  `[inherit a b; d c]` changes the order but that is minor.

Reply:
Marking this as resolved now that the RFC is using the simpler sugaring you suggested, but happy to revisit and discuss further :)